File: LeetCode/populating_next_right_pointers_in_each_node.py
#!/usr/bin/env python
#coding: utf-8

# Definition for a  binary tree node
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None
        self.next = None

class Solution:
    def _con(self, child, parent):
        if not child:
            return

        if parent:
            if child == parent.left:
                child.next = parent.right
            else:
                child.next = parent.next.left if parent.next else None

        self._con(child.left, child)
        self._con(child.right, child)

    # @param root, a tree node
    # @return nothing
    def connect(self, root):
        self._con(root, None)
        

if __name__ == '__main__':
    n1 = TreeNode(1)
    n2 = TreeNode(2)
    n3 = TreeNode(3)
    n4 = TreeNode(4)
    n5 = TreeNode(5)
    n6 = TreeNode(6)
    n7 = TreeNode(7)
    n1.left = n2
    n1.right = n3
    n2.left = n4
    n2.right = n5
    n3.left = n6
    n3.right = n7

    s = Solution()
    s.connect(n1)

    assert n1.next == None
    assert n2.next == n3
    assert n3.next == None
    assert n4.next == n5
    assert n5.next == n6
    assert n6.next == n7
    assert n7.next == None

File: LeetCode2/Solution/src/ios/ac/cn/PopulatingNextRightPointersInEachNode.java
package ios.ac.cn;

class TreeLinkNode {
	int val;
	TreeLinkNode left, right, next;

	TreeLinkNode(int x) {
		val = x;
	}
}

/**
 * 
 * 1. 注意要求用常量空间，这意味着不能用递归，如果用迭代，也不能用栈， 否则，不能算是用了常量空间
 * 
 * 2. 基本思想：每次迭代开始时，本层的 next 构成一个链表， 在此链表上遍历，为下一层建立 next 链表，这样，就使用常量空间 完成对树的遍历。
 * 
 */
public class PopulatingNextRightPointersInEachNode {
	public void connect(TreeLinkNode root) {
		TreeLinkNode start = root;

		while (start != null) {
			TreeLinkNode cur = start;
			while (cur != null) {
				if (cur.left != null) {
					cur.left.next = cur.right;
				}
				if (cur.right != null && cur.next != null) {
					cur.right.next = cur.next.left;
				}
				cur = cur.next;
			}
			start = start.left;
		}
	}
}
